### PR TITLE
index: add option "exclude_null" to index part definition

### LINKED ISSUE #4480: simplified partial index support
The proposal is to avoid lua function declaration and further calls in some common partial index cases, e.g. to define an index for not-null values only or for some limited subset of values. This could be achieved by means of additional options within index definition:
```
parts = {
    {1, 'integer', exc_values = {box.NULL}},
    {2, 'integer', inc_values = {1,2,3, box.NULL}}
},
dependences = {  -- not a part of the index
    {3, 'boolean', exc_values = {false}}
}
```

### PR DESCRIPTION
**Pull request by mary3000**
New option "exclude_null=true/false" was introduced.
"is_nullable=true" option will be set automatically.
When it is on, index skips tuples with "null" value of this part.
E.g. index `s:create_index('sk', {parts={{2, 'number', exclude_null=true}}})`
will ignore `{1, null}, {2, null}`, but will not ignore `{null, 1}, {1, 1}`.

Closes #4480

@TarantoolBot document
Title: new option "exclude_null=true/false"
Added new option for index part definition,
that allows index to skip tuples with null at this part.

### CODE REVIEW

File: src/box/lua/schema.lua [hunk 1]
```diff
@@ -775,6 +783,11 @@ local function update_index_parts(format, parts)
             box.error(box.error.ILLEGAL_PARAMS,
                       "options.parts[" .. i .. "]: type (boolean) is expected")
         end
+        if (not part.is_nullable) and part.exclude_null then
```
Comment (alyapunov) [inline src/box/lua/schema.lua:786]: I think we should made a bit better.
It a user sets `exclude_null=true` and doesn't set `is_nullable` we should internally set `is_nullable=true`.
Only when a user sets `exclude_null=true` and `is_nullable=false` it must be a error.

File: src/box/index.cc [hunk 2]
```diff
@@ -549,6 +549,28 @@ index_build(struct index *index, struct index *pk)
...
+	struct key_part* parts = key_def->parts;
+	for (uint32_t i = 0; i < key_def->part_count; ++i) {
+		if (!parts[i].exclude_null)
+			continue;
+		const char* field = tuple_field(*new_tuple, parts[i].fieldno);
+		if (mp_typeof(*field) == MP_NIL) {
+			if (old_tuple && *old_tuple != nullptr) {
+				*new_tuple = nullptr;
+				return false;
+			} else {
```
Comment (alyapunov) [inline src/box/index.cc:566]: What if old_tuple was excluded? As I see in this case we are deleting it from tree.. It does not actually delete anything because we didn't insert it, but it looks buggy.

**Review by alyapunov — CHANGES_REQUESTED** [review]
I left some comments. The main problem is `index_tuple_is_excluded` method is not optimal and is not obvious. That's a design problem, so it must be redesigned.
I also found a bug:
<details>
<summary>Reproduce</summary>

```lua
my.create_space{engine = 'memtx'}
my.create_index{type='tree', parts={{1, 'uint'}}}
my.create_index{type='tree', parts={{2, 'uint', exclude_null=true, is_nullable=true}}}

tarantool> s:replace{1}
Segmentation fault
  code: SEGV_MAPERR
  addr: (nil)
  context: 0x7f0ad19ff4c0
  siginfo: 0x7f0ad19ff5f0
  rax      0x0                0
  rbx      0x416ec538         1097778488
  rcx      0xffffffff         4294967295
  rdx      0xfffffffffffffffc -4
  rsi      0xffffffff         4294967295
  rdi      0x7f0a5100008e     139682285355150
  rsp      0x7f0ad19ffa70     139684443322992
  rbp      0x7f0ad19ffac0     139684443323072
  r8       0x0                0
  r9       0x0                0
  r10      0x0                0
  r11      0x202              514
  r12      0x0                0
  r13      0x0                0
  r14      0x416ed080         1097781376
  r15      0x549838           5543992
  rip      0x4525e7           4531687
  eflags   0x10206            66054
  cs       0x33               51
  gs       0x0                0
  fs       0x0                0
  cr2      0x0                0
  err      0x4                4
  oldmask  0x0                0
  trapno   0xe                14
Current time: 1607081698
Please file a bug at http://github.com/tarantool/tarantool/issues
Attempting backtrace... Note: since the server has already crashed, 
this may fail as well
#0  0x645ef9 in print_backtrace+9
#1  0x44d052 in _ZL12sig_fatal_cbiP9siginfo_tPv+1cb
#2  0x7f0ad4ab6390 in __restore_rt+0
#3  0x4525e7 in index_tuple_is_excluded+bc
#4  0x56c38f in index_replace+32
#5  0x56d23f in memtx_space_replace_all_keys+1cc
#6  0x56d3ed in memtx_space_execute_replace+d1
#7  0x45d91d in space_execute_dml+126
#8  0x488581 in box_process_rw+105
#9  0x48c80f in box_process1+8b
#10 0x48ccff in box_replace+82
#11 0x5498d8 in lbox_replace+a0
#12 0x6d410b in lj_BC_FUNCC+34
#13 0x6973a1 in lua_pcall+314
#14 0x621509 in luaT_call+29
#15 0x61862b in lua_main+b3
#16 0x618af8 in run_script_f+4b4
#17 0x44c9b3 in _ZL16fiber_cxx_invokePFiP13__va_list_tagES0_+1e
#18 0x642260 in fiber_loop+7c
#19 0x8bf0a1 in coro_init+4c
#20 (nil) in +4c
Aborted (core dumped)
```
</details>

File: src/box/key_def.h [hunk 3]
```diff
@@ -418,9 +423,10 @@ box_key_def_new(uint32_t *fields, uint32_t *types, uint32_t part_count);
  *
  * Default flag values are the following:
  *
- *  | Flag                         | Default value |
- *  | ---------------------------- | ------------- |
- *  | BOX_KEY_PART_DEF_IS_NULLABLE | 0 (unset)     |
+ *  | Flag                          | Default value |
+ *  | ----------------------------- | ------------- |
+ *  | BOX_KEY_PART_DEF_IS_NULLABLE  | 0 (unset)     |
+ *  | BOX_KEY_PART_DEF_EXCLUDE_NULL | 0 (unset)     |
```
Comment (Totktonada) [inline src/box/key_def.h:429]: When the new flag will land to master, please, don't forget to request to document the new changes. This particular API is not documented yet, so you can just leave a note in https://github.com/tarantool/doc/issues/1611. Or you can add it to the docbot request in the commit message.

BTW, it should be `@TarantoolBot document`, not `@documentbot document` if I don't miss something.

File: src/box/lua/key_def.c [hunk 4]
```diff
@@ -62,6 +62,9 @@ luaT_push_key_def(struct lua_State *L, const struct key_def *key_def)
 		lua_pushboolean(L, key_part_is_nullable(part));
 		lua_setfield(L, -2, "is_nullable");
 
+		lua_pushboolean(L, part->exclude_null);
+		lua_setfield(L, -2, "exclude_null");
```
Comment (Totktonada) [inline src/box/lua/key_def.c:66]: We want to keep the [tuple-keydef](https://github.com/tarantool/tuple-keydef) module compatible with the built-in one, so, please, file an issue to its repository, when the change will land to master (or after agreement about the future Lua API). It is nice that you updated the public C API: so we can support it in the external module as well.

File: test/app-tap/module_api.c [hunk 5]
```diff
@@ -1075,7 +1075,7 @@ test_key_def_merge(struct lua_State *L)
 
 	/* Non-conventional prerequisite: list of known flags. */
 	uint32_t known_flags = key_part_def_known_flags();
-	assert(known_flags == BOX_KEY_PART_DEF_IS_NULLABLE);
+	assert(known_flags == (BOX_KEY_PART_DEF_IS_NULLABLE | BOX_KEY_PART_DEF_EXCLUDE_NULL));
```
Comment (Totktonada) [inline test/app-tap/module_api.c:1078]: Please, update also the assert below about default values of flags. Otherwise direct initialization of the `flags` field with zero in the test cases below will not be a kind of 'legal'.

File: src/box/lua/key_def.c (hunk 4 above)
Comment (Totktonada) [inline src/box/lua/key_def.c:66]: BTW, can we don't push the `exclude_null` field when it has the default value? I see, you follows  the behaviour around `is_nullable`, but it looks as the bad pattern for me. We don't show default collation or JSON path. We don't show default `is_nullable = false` in a space format serialization. However we show it in the index parts serialization (and I don't know why).

@alyapunov What do you think?

File: test/app-tap/module_api.c (hunk 5 above)
Comment (Totktonada) [inline test/app-tap/module_api.c:1078]: It is maybe a bit too formal, but we have test cases for merging (`box_key_def_merge()`) key parts, which differs only in the `is_nullable` flag (it is case 6 and 7). It looks logical to have the similar ones for the new `exclude_null` flag.

File: src/box/lua/key_def.c (hunk 4 above)
Comment (alyapunov) [reply src/box/lua/key_def.c:66]: > it looks as the bad pattern for me
I agree. If we also remove default `is_nullable = false` from index - I think it will be perfect.
PS I also thought about it but `is_nullable` implementation confused me.

File: test/app-tap/module_api.c (hunk 5 above)
Comment (alyapunov) [reply test/app-tap/module_api.c]: Agree, more test cases is better.

File: test/box-tap/key_def.test.lua [hunk 6]
```diff
@@ -34,9 +35,6 @@ local function set_key_part_defaults(parts)
     local res = {}
     for i, part in ipairs(parts) do
         res[i] = table.copy(part)
-        if res[i].is_nullable == nil then
-            res[i].is_nullable = false
-        end
```
Comment (Totktonada) [inline test/box-tap/key_def.test.lua:39]: It seems, the whole function is not needed anymore.

File: test/app-tap/module_api.c [hunk 7]
```diff
@@ -1075,13 +1075,15 @@ test_key_def_merge(struct lua_State *L)
 
 	/* Non-conventional prerequisite: list of known flags. */
 	uint32_t known_flags = key_part_def_known_flags();
-	assert(known_flags == BOX_KEY_PART_DEF_IS_NULLABLE);
+	assert(known_flags == (BOX_KEY_PART_DEF_IS_NULLABLE |
+				BOX_KEY_PART_DEF_EXCLUDE_NULL));
```
Comment (Totktonada) [inline test/app-tap/module_api.c:1079]: Indent looks broken here.

File: test/app-tap/module_api.c [hunk 8]
```diff
@@ -1645,6 +1647,48 @@ test_key_def_merge(struct lua_State *L)
 	key_def_check_merge(a_25, lengthof(a_25), b_25, lengthof(b_25),
 			    exp_25, lengthof(exp_25));
 
+	/*
+	* Case 26: exclude_null=true + exclude_null=false.
+	*
+	* Interpretation: the same as for the case 3.
+	*/
```
Comment (Totktonada) [inline test/app-tap/module_api.c:1654]: (Sorry if I comment too early and it is just draft. I got email regarding the changes and so glance on them.)

Multiline comments are formatted this way:

```
/*
 * foo
 * bar
 */
```

Linus regarding this: https://lkml.org/lkml/2016/7/8/625

**Review by alyapunov — CHANGES_REQUESTED** [review]
There's one significant comment.
Please squash commits.
I didn't found a test case that checks that it's not possible to create primary index with `exclude_null`.

File: src/box/key_def.c [hunk 9]
```diff
@@ -808,6 +831,11 @@ key_def_encode_parts(char *data, const struct key_part_def *parts,
 					     strlen(PART_OPT_NULLABILITY));
 			data = mp_encode_bool(data, part->is_nullable);
 		}
+		if (part->exclude_null) {
+			data = mp_encode_str(data, PART_OPT_EXCLUDE_NULL,
+								 strlen(PART_OPT_EXCLUDE_NULL));
```
Comment (alyapunov) [inline src/box/key_def.c:836]: What a strange indent?

File: src/box/key_def.c [hunk 10]
```diff
@@ -1070,6 +1099,8 @@ key_def_merge(const struct key_def *first, const struct key_def *second)
 	new_def->part_count = new_part_count;
 	new_def->unique_part_count = new_part_count;
 	new_def->is_nullable = first->is_nullable || second->is_nullable;
+	new_def->has_exclude_null = first->has_exclude_null ||
+				      second->has_exclude_null;
```
Comment (alyapunov) [inline src/box/key_def.c:1103]: Please check the indent too

**Review by alyapunov — APPROVED** [review]
LGTM

**Review by Korablev77 — COMMENTED** [review]
Write doc request as you want to see it in documentation: for instance describe how it interacts with is_nullbable fields (you did it in commit message, but it doesn't get into doc request). What else might be useful: underline that index does not store filtered tuples; add lines about on_replace triggers - whether they fire during insertion of skipped tuple or not; can this setting be changed dynamically (after filling the index); default per index value for this option.
/* Skip comment below, it is not relevant. */
Concerning implementation: why do you have to *filter* tuples with this option? Contrary, you can avoid adding them to the corresponding index. For sure, you approach is likely to be easier in its implementation, but what's the point of storing keys that never will be shown? It reduces seek overhead and saves some space especially for vinyl engine.
Upd.: oops, sorry, I leave this comment before dived into the patch. Tuples actually are filtered before insertion to the index. Please skip the second part of comment.

File: src/box/index.cc [hunk 11]
```diff
@@ -554,6 +554,23 @@ index_build(struct index *index, struct index *pk)
 	return 0;
 }
 
+struct tuple *
+get_tuple_for_index_slow(struct index *index, struct tuple *tuple)
```
Comment (Korablev77) [inline src/box/index.cc:558]: Don't like the name of this function. We name function starting from the subject: index_get_tuple_slow().
Moreover, this function works as a predicate, so I'd refactor it this way:
 - make it return boolean (false, if tuple doesn't satisfy exclude_null condition, true otherwise)
 - rename to sort of index_tuple_is_excluded/filtered.

File: src/box/index.h [hunk 12]
```diff
@@ -623,11 +623,36 @@ index_get(struct index *index, const char *key,
 	return index->vtab->get(index, key, part_count, result);
 }
 
+/**
+ * Get tuple to be inserted in index, based on index-specific constraints
+ * (current constraint: if exclude_null = true, return NULL)
+ * @param index
+ * @param tuple
```
Comment (Korablev77) [inline src/box/index.h:630]: I'd better remove this uninformative doxygen comments (or add smth meaningful).

File: src/box/lua/key_def.c [hunk 13]
```diff
@@ -59,8 +59,15 @@ luaT_push_key_def(struct lua_State *L, const struct key_def *key_def)
 			lua_setfield(L, -2, "path");
 		}
 
-		lua_pushboolean(L, key_part_is_nullable(part));
-		lua_setfield(L, -2, "is_nullable");
+		if (key_part_is_nullable(part)) {
+			lua_pushboolean(L, true);
+			lua_setfield(L, -2, "is_nullable");
```
Comment (Korablev77) [inline src/box/lua/key_def.c:64]: Can we avoid changing this behaviour in current patch? I guess it is okay to hide is_nullable option, but could you please move this to a separate patch?

File: src/box/lua/schema.lua (hunk 1 above)
Comment (Korablev77) [reply src/box/lua/schema.lua:796]: @alyapunov Why it should result in error? I believe it is OK: is_nullable field can't contain nulls, so exclude_null filter just does nothing.

File: src/box/lua/schema.lua [hunk 14]
```diff
@@ -785,6 +793,14 @@ local function update_index_parts(format, parts)
...
+            if part.is_nullable ~= nil then
+                box.error(box.error.ILLEGAL_PARAMS,
+                       "options.parts[" .. i .. "]: exclude_null=true and " ..
+                       "is_nullable=false are contradictory")
```
Comment (Korablev77) [inline src/box/lua/schema.lua:800]: I'd say 'incompatible' (still personally I believe that they should be compatible).

File: test/engine/null.test.lua [hunk 15]
```diff
@@ -612,3 +612,117 @@ s:insert{1, -1, 1}
...
+s = box.schema.space.create('test')
+pk = s:create_index('pk', {type='tree', parts={{1, 'uint'}}})
+sk = s:create_index('sk', {type='tree', parts={{2, 'uint', exclude_null=true}}})
+s:replace{1}
+sk:select{}
+s:select{}
+
+-- Forbid creating primary key with exclude_null option
+s:drop()
+s = box.schema.space.create('test', {engine=engine})
```
Comment (Korablev77) [inline test/engine/null.test.lua:727]: Forgot to drop s..
Also I would check that after restoring from snapshot exclude_null still works.

**Comment by mary3000** [comment]
> Write doc request as you want to see it in documentation: for instance describe how it interacts with is_nullbable fields (you did it in commit message, but it doesn't get into doc request).

I used TarantoolBot in the commit message, I thought it will create doc request automatically after closing the issue?

File: src/box/index.cc (hunk 11 above)
Comment (mary3000) [reply src/box/index.cc:558]: It was a predicate in my first version. But then I need to do two different things: 1) evaluate predicate 2) change tuple to null if the predicate is true. If I do this within one function, then it complicates its behavior (see [comment](https://github.com/tarantool/tarantool/pull/5481#discussion_r535891800) from @alyapunov). If I set null value manually, then there will be a lot of repeating code in different places which I don't like. So I decided that the meaning of this function should be only smth like "return tuple which will be actually inserted in index". So no predicate evaluation is needed.

File: src/box/lua/key_def.c (hunk 13 above)
Comment (mary3000) [reply src/box/lua/key_def.c]: fixed

File: src/box/index.h (hunk 12 above)
Comment (mary3000) [reply src/box/index.h]: fixed

File: src/box/lua/schema.lua (hunk 14 above)
Comment (mary3000) [reply src/box/lua/schema.lua]: fixed

File: test/engine/null.test.lua (hunk 15 above)
Comment (mary3000) [reply test/engine/null.test.lua:727]: Thx, good point! I added test, found a bug with memtx engine, so now yet another check is added (see [memtx_tree_index_build_next](https://github.com/tarantool/tarantool/pull/5481/files#diff-fb2f7ff1cd0f93acca771e67cfce349c6a4ebc4b1a602d59667aa1d0cc216a8fR1298)). Initially I thought in memtx every change should be passed through `index_replace` function.

File: src/box/lua/schema.lua (hunk 1 above)
Comment (mary3000) [reply src/box/lua/schema.lua:796]: I write some thoughts here: now because `exclude_null` turns on `is_nullable`, it implies that `exclude_null` cannot be set in primary key. That means tuple is inserted at least in one index. But if we want to support setting both `exclude_null=true` and `is_nullable=false`, then I suppose primary key should be allowed to have the new option. And then we should think about `on_replace` triggers behavior as Nikita already mentioned.

**Comment by mary3000** [comment]
> add lines about on_replace triggers - whether they fire during insertion of skipped tuple or not; 

@Korablev77 now because primary key can't have this option turned on, I guess there is nothing to say about on_replace - tuple will get into primary key anyway.

File: src/box/index.cc (hunk 11 above)
Comment (alyapunov) [reply src/box/index.cc]: why not `struct tuple* index_filter_tuple_slow(struct index *index, struct tuple *tuple)`?

File: src/box/lua/schema.lua (hunk 1 above)
Comment (Korablev77) [reply src/box/lua/schema.lua:796]: I discussed the problem with Alexander. The thing is if we allowed this configuration (exclude_null=true and is_nullable=false) we would face the 'race' when inserting null to the index. One the one hand, exclude_null throws away tuple containing null (i.e. no error is raised); on the other hand, is_nullable property should raise an error (since we are inserting incompatible value). So depending on which check is fired first, the error may be raised. So to avoid this mess Alexander suggests to make these options incompatible. After this explanation I agree with his point and propose to leave it as is for now.

File: src/box/index.cc (hunk 11 above)
Comment (Korablev77) [reply src/box/index.cc]: Why do you have to change tuple to null? Using current schema anyway two steps are required:
1) invoke get_tuple_for_index_slow()
2) return/skip tuple if the result is null
Using predicate:
1) invoke index_tuple_is_filtered()
2) return/skip tuple if the result is true

File: src/box/vinyl.c [hunk 16]
```diff
@@ -1946,8 +1950,9 @@ vy_insert_first_upsert(struct vy_env *env, struct vy_tx *tx,
 	if (vy_tx_set(tx, pk, stmt) != 0)
 		return -1;
 	for (uint32_t i = 1; i < space->index_count; ++i) {
+		struct tuple *new_tuple = index_get_tuple(space->index[i], stmt);
```
Comment (Korablev77) [inline src/box/vinyl.c:1953]: Would be nice to have an assertion (new_tuple != NULL) here.

Comment (Korablev77) [reply src/box/vinyl.c:1953]: Oh, it actually may cause a crash:
```
s = box.schema.space.create('test', { engine = 'vinyl'})
_ = s:create_index('pk')
sk = s:create_index('sk', {parts={{2, 'number', exclude_null=true}}})

s:replace{1, 1}
box.snapshot()
s:upsert({2, box.NULL}, {{'=', 2, box.NULL}})

#0  0x55e89080b582 in print_backtrace+9
#1  0x55e890604070 in _ZL12sig_fatal_cbiP9siginfo_tPv+1e7
#2  0x7f3d8373c890 in funlockfile+50
#3  0x55e890767505 in tuple_format+10
#4  0x55e8907678b3 in vy_stmt_is_key+18
#5  0x55e8907679aa in vy_stmt_hint+1c
#6  0x55e8907712b2 in vy_tx_set+a2
#7  0x55e89073eb2d in vy_insert_first_upsert+1ab
#8  0x55e89073f415 in vy_upsert+325
#9  0x55e89073fd8a in vinyl_space_execute_upsert+69
#10 0x55e890614ff6 in space_execute_dml+211
#11 0x55e8906408f6 in box_process_rw+10b
#12 0x55e890645434 in box_process1+172
#13 0x55e890645b8b in box_upsert+c4
#14 0x55e890709805 in lbox_upsert+131
#15 0x55e89089d7cb in lj_BC_FUNCC+34
#16 0x55e89085f8b4 in lua_pcall+332
#17 0x55e8907e53a6 in luaT_call+29

```

**Comment by Korablev77** [comment]
Check also vy_build_on_replace() function: there's no index_get_tuple() call however it still inserts tuple into tx write set.

**Comment by Korablev77** [comment]
Another one tricky place: vy_build_recover_stmt(). There we insert statement right to L0 passing tx write set. Please, check it, I guess here we need also verify the tuple.

File: src/box/vinyl.c [hunk 17]
```diff
@@ -4081,6 +4098,8 @@ vy_build_recover_stmt(struct vy_lsm *lsm, struct vy_lsm *pk,
 static int
 vy_build_recover_mem(struct vy_lsm *lsm, struct vy_lsm *pk, struct vy_mem *mem)
 {
+	//assert(0);
+	say_info("vy_build_recover_mem");
```
Comment (Korablev77) [inline src/box/vinyl.c:4102]: Artifacts after debugging...

File: src/box/lua/key_def.c [hunk 18]
```diff
@@ -434,6 +446,7 @@ lbox_key_def_new(struct lua_State *L)
 		return luaL_error(L, "Bad params, use: key_def.new({"
 				  "{fieldno = fieldno, type = type"
 				  "[, is_nullable = <boolean>]"
+	  			  "[, exclude_null = <boolean>]"
```
Comment (Korablev77) [inline src/box/lua/key_def.c:449]: Spaces between tabs

File: test/vinyl/errinj_ddl.test.lua [hunk 19]
```diff
@@ -57,6 +57,33 @@ s.index.sk:stat().memory.rows
...
 s:drop()
 
+-- exclude_null: correct recovering tuples from memory (vy_build_recover_stmt)
+s = box.schema.space.create('test', {engine='vinyl'})
+_ = s:create_index('pk')
+
+s:replace{1, 1}
+s:replace{2, 2}
+
+errinj.set("ERRINJ_VY_RUN_WRITE_DELAY", true)
```
Comment (Korablev77) [inline test/vinyl/errinj_ddl.test.lua:67]: Nice test, thx

File: test/sql/misc.test.lua [hunk 20]
```diff
@@ -79,3 +79,30 @@ box.execute('SELECT a;')
 diag = box.error.last()
 box.execute('SELECT * FROM (VALUES(true));')
 diag == box.error.last()
+
+-- exclude_null + SQL correctness
+box.execute([[CREATE TABLE j (s1 INT PRIMARY KEY, s2 STRING, s3 VARBINARY) WITH ENGINE='vinyl';]])
```
Comment (Korablev77) [inline test/sql/misc.test.lua:84]: sql/ suit is ran using both engines: at the start default engine is set
`_ = box.space._session_settings:update('sql_default_engine', {{'=', 2, engine}})
`
So you don't have to specify vinyl engine.

**Comment by Korablev77** [comment]
SQL tests are failing due to assertion fault:
`tarantool: /tarantool/src/box/sql/where.c:3953: wherePathSolver: Assertion `nFrom != 0' failed.
`

File: src/box/sql/where.c [hunk 21]
```diff
@@ -2926,6 +2926,9 @@ whereLoopAddBtree(WhereLoopBuilder * pBuilder,	/* WHERE clause information */
 	for (uint32_t i = 0; i < idx_count; iSortIdx++, i++) {
 		if (i > 0)
 			probe = space->index[i]->def;
+		/* Such index may possibly contain not all tuples, so skip it */
+		if (pSrc->pIBIndex == NULL && probe->key_def->has_exclude_null)
```
Comment (Korablev77) [inline src/box/sql/where.c:2930]: This makes indexes with exclude_null always invisible in SQL, although they can be used. For instance (schema is taken from your test):
```
explain query plan select s2 from j where s2 == 3;
[0, 0, 0, 'SCAN TABLE J (~262144 rows)']
```
If index didn't contain exclude_null option, plan would be following:
`
[0, 0, 0, 'SEARCH TABLE J USING COVERING INDEX I3 (S2=?) (~1 row)']
`
I guess we should come for a bit more sophisticated heuristics (probably it should be implemented not in scope of this issue). Please, find out if we can fix this easily and in case we can't - file an issue.

Comment (mary3000) [reply src/box/sql/where.c:2931]: In any case, it is better to do in a new issue - I created it [here](https://github.com/tarantool/tarantool/issues/5718#issue-787979259)

File: test/sql/misc.test.lua [hunk 22]
```diff
@@ -79,3 +79,29 @@ box.execute('SELECT a;')
...
+box.execute([[UPDATE j INDEXED BY i3 SET s2 = NULL;]])
+box.execute([[INSERT INTO j VALUES (3, 'a', X'33');]])
+
+box.execute([[SELECT * FROM j;]])
+box.execute([[SELECT * FROM j INDEXED BY I3;]])
+
+box.execute([[UPDATE j INDEXED BY i3 SET s3 = NULL;]])
+s:select{}
+-- Fixme: exclude_null should be true
+box.space.J.index.I3:alter({parts={2,exclude_null=true}})
```
Comment (Korablev77) [inline test/sql/misc.test.lua:104]: Didn't get this test...Am I correct that you intended to shown that missing type specification in key definition results in re-setting exclude_null property to false? Can you fix it? Is it related to current PR (https://github.com/tarantool/tarantool/pull/5731)? If so, could you please move this case out of sql suite?

**Comment by Korablev77** [comment]
I left one tiny comment concerning ignored exclude_null option when key part type is missing. Otherwise LGTM.
Also I'd add to commit following table:
| Nullable/Exclude NULL | False | True |
| ----------------------- | ---------- | -------------- |
| False     |  +     | +      |
| True       | - *   |  + **     |

`*` is not allowed
`**` Partial index

**Review by Korablev77 — APPROVED** [review]

**Comment by Totktonada** [comment]
> | Nullable/Exclude NULL | False | True |
> | ----------------------- | ---------- | -------------- |
> | False     |  +     | +      |
> | True       | - *   |  + **     |
>
> `*` is not allowed
> `**` Partial index

I guess this table should be transposed. `{nullable = true, exclude_null = false}` is okay (usual index), but `{nullable = false, exclude_null = true}` has no sense and I guess it is forbidden. Right?

File: test/sql/misc.test.lua (hunk 22 above)
Comment (mary3000) [reply test/sql/misc.test.lua:104]: Yes? it's about this bug. I'll remove this test case from this pull request then.

**Comment by Korablev77** [comment]
@Totktonada Yep, for sure, I've updated table.

**Review by alyapunov — APPROVED** [review]
LGTM

**Comment by Korablev77** [comment]
Pushed to master.
